myst_nbs/case_studies/wrapping_jax_function.myst.md

```diff
@@ -0,0 +1,775 @@
+pycharm:
+  name: '#%%
+
+    '
+---
+jitted_hmm_logp_value_and_grad = jax.jit(jax.value_and_grad(vec_hmm_logp, argnums=list(range(5))))
+```
+
+```{code-cell} ipython3
+class HmmLogpValueGradOp(Op):
+    # By default only return the first output
+    default_output = 0
+
+    def make_node(self, *inputs):
+        inputs = [at.as_tensor_variable(inp) for inp in inputs]
+        # We now have one output for the function value, and one output for each gradient
+        outputs = [at.dscalar()] + [inp.type() for inp in inputs]
+        return Apply(self, inputs, outputs)
+
+    def perform(self, node, inputs, outputs):
+        result, grad_results = jitted_hmm_logp_value_and_grad(*inputs)
+        outputs[0][0] = np.asarray(result, dtype=node.outputs[0].dtype)
+        for i, grad_result in enumerate(grad_results, start=1):
+            outputs[i][0] = np.asarray(grad_result, dtype=node.outputs[i].dtype)
+
+    def grad(self, inputs, output_gradients):
+        # The `Op` computes its own gradients, so we call it again.
+        value = self(*inputs)
+        # We hid the gradient outputs by setting `default_update=0`, but we
+        # can retrieve them anytime by accessing the `Apply` node via `value.owner`
+        gradients = value.owner.outputs[1:]
+
+        # Make sure the user is not trying to take the gradient with respect to
+        # the gradient outputs! That would require computing the second order
+        # derivatives
+        assert all(
+            isinstance(g.type, aesara.gradient.DisconnectedType) for g in output_gradients[1:]
+        )
+
+        return [output_gradients[0] * grad for grad in gradients]
+
+
+hmm_logp_value_grad_op = HmmLogpValueGradOp()
+```
+
+We check again that we can take the gradient using Aesara `grad` interface
+
+```{code-cell} ipython3
+emission_signal_variable = at.as_tensor_variable(emission_signal_true)
+# Only the first output is assigned to the variable `x`, due to `default_output=0`
+x = hmm_logp_value_grad_op(
+    emission_observed,
+    emission_signal_variable,
+    emission_noise_true,
+    logp_initial_state_true,
+    logp_transition_true,
+)
+at.grad(x, emission_signal_variable).eval()
+```
+
+(aesara_vs_jax)=
+## Notes on using Aesara vs JAX
+
++++
+
+### When should you write functions in JAX?
+
++++
+
+As we have seen, it is pretty straightforward to interface between Aesara graphs and JAX functions. 
+
+This can be very handy when you want to combine previously implemented JAX function with PyMC models. We used a marginalized HMM log-likelihood in this example, but the same strategy could be used to do Bayesian inference with Deep Neural Networks or Differential Equations, or pretty much any other functions implemented in JAX that can be used in the context of a Bayesian model.
+
+It can also be important, if you need to make use of JAX's unique features like vectorization, support for tree structures, or make use of its fine-grained parallelization, or GPU and TPU capabilities.
```

Comment:
```suggestion
It can also be worth it, if you need to make use of JAX's unique features like vectorization, support for tree structures, or its fine-grained parallelization, and GPU and TPU capabilities.
```